private/buf/cmd/buf/command/decode/decode.go

```diff
@@ -0,0 +1,171 @@
+			func(ctx context.Context, container appflag.Container) error {
+				return run(ctx, container, flags)
+			},
+			bufcli.NewErrorInterceptor(),
+		),
+		BindFlags: flags.Bind,
+	}
+}
+
+type flags struct {
+	ErrorFormat string
+	Source      string
+	Type        string
+	Output      string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.ErrorFormat,
+		errorFormatFlagName,
+		"text",
+		fmt.Sprintf(
+			"The format for build errors, printed to stderr. Must be one of %s.",
+			stringutil.SliceToString(bufanalysis.AllFormatStrings),
+		),
+	)
+	flagSet.StringVar(
+		&f.Source,
+		sourceFlagName,
+		"",
+		"The source that defines the serialized descriptor (e.g. buf.build/acme/weather)",
+	)
+	flagSet.StringVar(
+		&f.Type,
+		typeFlagName,
+		"",
+		"The fully-qualified type name of the serialized descriptor (e.g. acme.weather.v1.Units)",
+	)
+	flagSet.StringVarP(
+		&f.Output,
+		outputFlagName,
+		outputFlagShortName,
+		"-",
+		// TODO: If we ever support other formats (e.g. prototext), we will need
+		// to build a buffetch.ProtoEncodingRefParser.
+		`The location to write the decoded result to. This is always JSON for now.`,
+	)
+}
+
+func run(
+	ctx context.Context,
+	container appflag.Container,
+	flags *flags,
+) error {
+	if err := bufcli.ValidateErrorFormatFlag(flags.ErrorFormat, errorFormatFlagName); err != nil {
+		return err
+	}
+	descriptorBytes, err := io.ReadAll(container.Stdin())
+	if err != nil {
+		return err
+	}
+	if len(descriptorBytes) == 0 {
+		return fmt.Errorf("stdin is required as the input")
+	}
+	protoSource, protoType, err := parseSourceAndType(ctx, flags.Source, flags.Type)
+	if err != nil {
+		return err
+	}
+	registryProvider, err := bufcli.NewRegistryProvider(ctx, container)
+	if err != nil {
+		return err
+	}
+	image, err := bufcli.NewImageForSource(ctx, container, registryProvider, protoSource, flags.ErrorFormat)
+	if err != nil {
+		return err
+	}
+	message, err := bufreflect.NewMessage(ctx, image, protoType)
+	if err != nil {
+		return err
+	}
+	if err := protoencoding.NewWireUnmarshaler(nil).Unmarshal(descriptorBytes, message); err != nil {
+		return err
+	}
+	return bufcli.NewWireProtoEncodingWriter(
+		container.Logger(),
+	).PutMessage(
+		ctx,
+		container,
+		message,
+		flags.Output,
+	)
+}
+
+// parseSourceAndType returns the moduleReference and typeName from the source and type provided by the user.
+// When source is not provided, we assume the type is a fully-qualified path to the type and try to parse it.
+// When source and type are redundantly specified (e.g. buf.build/acme/weather and buf.build/acme/weather#weather.v1.Units),
+// we are not handling it and just let it fail.
+func parseSourceAndType(
+	ctx context.Context,
+	flagSource string,
+	flagType string,
+) (moduleReference string, typeName string, _ error) {
+	if flagSource != "" && flagType != "" {
+		return flagSource, flagType, nil
+	}
```

Comment:
We should still be validating the `flagType` here according to the `ValidateTypeName` helper. We validate it in `bufreflect.ParseFullyQualifiedPath`, so we should make the same validation guarantee for all cases in this function.

Reply:
sounds good to me! I am thinking about putting `parseSourceAndType`, `ParseFullyQualifiedPath` and `validateTypeName` together under `buf decode` for now, and move them to `bufreflect` when we need to use them in `buf encode`, so that `parseSourceAndType` can use `validateTypeName`